keps/sig-apps/0033-20190316-app-pods-graceful-update.md

```diff
@@ -0,0 +1,191 @@
+
+
+### Goals
+- For `StatefulSet` and `DaemonSet`, patch Pods by restarting containers locally when **only** main container (`spec.Containers`) images changed
+- TODO: make `Deployment` graceful pod update into Goals?
+
+### NonGoals
+- `Deployment` graceful Pod update. I put this as non-goals for the following reasons:
+    - Based on current implementation, it's hard to implement Pod patching cleanly while keeping current behavior without major refactoring / redesign
+    - Stateless services has less startup cost, and since we create new replica first, availability is less likely to get affected
+- Graceful update Pod in senarios other than image change, i.e. environment variable change, local volume mount change, etc
+    - These scenarios are less likely to happen and currently we don't allow upate `Pod` fields other than images, activeDeadlineSeconds, and tolerations
+
+
+## Proposal
+
+### User Stories
+- As an application developer, I don't want my application to be restarted unnecessarily just because other sidecar teams wants to roll out their changes,
+as crash handling routine such as leadership switch might affect availability
+- As a sidecar developer, I don't want to wait for application team's next release cycle to roll out my changes - some application teams has really long 
+release cycles and that increases the number of versions we need to maintain
+- As a engineering manager, I don't want to migrate to k8s as my current VM environment can support separate container deployment but k8s does not, which
+is a regression to me
+- As a infrastructure developer, I don't want to get paged about cluster churns such as etcd / k8s api latency / failure spikes as currently there are
+a lot of api calls taking place during Pod restart
+
+### Implementation Details/Notes/Constraints
+Application updates are controlled by update strategy in their specs:
+- For `StatefulSet`, it is `StatefulSetUpdateStrategy`
+- For `DaemonSet`, it is `DaemonSetUpdateStrategy`
+These update strategies can be extended with an additional field describing how their pods should be updated.
+
+First we define a generic pod update strategy enumeration type, which can be applied to any application spec
+```go
+// PodUpdateStrategyType is a string enumeration type that enumerates
+// all possible ways we can update a Pod when updating application
+type PodUpdateStrategyType string
+
+const (
+    // PatchIfPossiblePodUpdateStrategyType indicates that we try to patch Pod instead of
+    // recreate Pod when possible. Currently we patch Pod only when any of the main
+    // containers (those in Spec.Containers) has image changes
+    PatchIfPossiblePodUpdateStrategyType = "PatchIfPossible"
+
+    // RecreatePodUpdateStrategyType indicates that we always delete Pod and create new Pod
+    // during Pod update, which is the current behavior
+    RecreatePodUpdateStrategyType = "ReCreate"
```

Comment:
I think "ReCreate" here is more appropriate than "Delete". 
On the other hand, "Delete" might be confusing because of the "OnDelete" StatefulSetUpdateStrategyType.

Reply:
Yep "DeletePodUpdateStrategyType" also sounds weird therefore I'm more inclined to "ReCreate". Currently we have no plan to support pod in place upgrade for deployment as it's upgrade mechanism is a bit different by using replica set.

Event we want to support this in deployment, we can put this in to `RollingUpdateDeployment` so it will only appear in rolling update and thus less confusion.

I can change the name from "ReCreate" to "Recreate" - just we are already using this as one word in k8s